File: java17/src/main/java/folio/codinginterview/application/usecase/order/NewOrderUsecase.java
package folio.codinginterview.application.usecase.order;

import folio.codinginterview.application.repository.AccountRepository;
import folio.codinginterview.application.repository.PortfolioRepository;
import folio.codinginterview.domain.Account;
import folio.codinginterview.domain.AppConstants;
import folio.codinginterview.domain.UserId;

import java.math.BigDecimal;
import java.util.concurrent.CompletableFuture;

public final class NewOrderUsecase {
    public record Input(UserId userId, BigDecimal amount) {}

    public static class Exception extends RuntimeException {
        protected Exception(String msg) { super(msg); }
    }

    public static final class UserAlreadyExists extends Exception {
        public static final UserAlreadyExists INSTANCE = new UserAlreadyExists();
        private UserAlreadyExists() { super("user already exists"); }
    }

    public static final class AmountTooSmall extends Exception {
        public static final AmountTooSmall INSTANCE = new AmountTooSmall();
        private AmountTooSmall() { super("amount too small"); }
    }

    private final AccountRepository accountRepository;
    private final PortfolioRepository portfolioRepository;

    public NewOrderUsecase(AccountRepository accountRepository, PortfolioRepository portfolioRepository) {
        this.accountRepository = accountRepository;
        this.portfolioRepository = portfolioRepository;
    }

    public CompletableFuture<Void> run(Input input) {
        if (input.amount().compareTo(AppConstants.MIN_OPERATION_AMOUNT) < 0) {
            CompletableFuture<Void> failed = new CompletableFuture<>();
            failed.completeExceptionally(AmountTooSmall.INSTANCE);
            return failed;
        }
        return accountRepository.exists(input.userId()).thenCompose(exists -> {
            if (exists) {
                CompletableFuture<Void> failed = new CompletableFuture<>();
                failed.completeExceptionally(UserAlreadyExists.INSTANCE);
                return failed;
            }
            return portfolioRepository.get().thenCompose(portfolio -> {
                var account = Account.openAccount(input.amount(), portfolio);
                return accountRepository.upsert(input.userId(), account);
            });
        });
    }
}

File: php/src/CodingInterview/Application/Usecase/Order/NewOrderUsecase.php
<?php

declare(strict_types=1);

namespace Folio\CodingInterview\Application\Usecase\Order;

use Folio\CodingInterview\Application\Repository\AccountRepository;
use Folio\CodingInterview\Application\Repository\PortfolioRepository;
use Folio\CodingInterview\Domain\Account;
use Folio\CodingInterview\Domain\AppConstants;
use Folio\CodingInterview\Domain\BigDecimal;
use Folio\CodingInterview\Domain\UserId;

final class NewOrderUsecaseInput
{
    public function __construct(
        public readonly UserId $userId,
        public readonly BigDecimal $amount,
    ) {}
}

final class NewOrderUserAlreadyExistsException extends \RuntimeException
{
    public function __construct() { parent::__construct('user already has account'); }
}

final class NewOrderAmountTooSmallException extends \RuntimeException
{
    public function __construct() { parent::__construct('amount is too small'); }
}

final class NewOrderUsecase
{
    public function __construct(
        private readonly AccountRepository $accountRepository,
        private readonly PortfolioRepository $portfolioRepository,
    ) {}

    public function run(NewOrderUsecaseInput $input): void
    {
        if ($input->amount->lt(AppConstants::minOperationAmount())) {
            throw new NewOrderAmountTooSmallException();
        }
        if ($this->accountRepository->exists($input->userId)) {
            throw new NewOrderUserAlreadyExistsException();
        }
        $portfolio = $this->portfolioRepository->get();
        $account = Account::openAccount($input->amount, $portfolio);
        $this->accountRepository->upsert($input->userId, $account);
    }
}
